### PR TITLE
Implement the function of OutScaleForTraining/OutScaleForInference in dygraph

### PR DESCRIPTION
**Pull request by gfwm2013**
### PR types
New features

### PR changes
 APIs

### Describe
Implement the function of OutScaleForTraining / OutScaleForInference in dygraph

### CODE REVIEW

File: paddle/fluid/pybind/op_function_generator.cc [hunk 1]
```diff
@@ -114,6 +116,8 @@ std::map<std::string, std::set<std::string>> op_passing_outs_map = {
     {"check_finite_and_unscale", {"Out", "FoundInfinite"}},
     {"update_loss_scaling",
      {"Out", "LossScaling", "OutGoodSteps", "OutBadSteps"}},
+    {"amp_check_finite_and_scale", {"Out", "FoundInfinite"}},
```
Comment (wzzju) [inline paddle/fluid/pybind/op_function_generator.cc]: ?? This is not needed.

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 2]
```diff
@@ -16,13 +16,13 @@
 import numpy as np
 import sys
 import paddle
-from paddle.fluid import dygraph
-from paddle.fluid.dygraph.nn import Conv2D
-from paddle.fluid.dygraph.nn import Linear
+from paddle.fluid import dygraph, core
+from paddle.fluid.dygraph.nn import Conv2D, Linear, BatchNorm, Pool2D, Conv2DTranspose, PRelu
+from paddle.nn.layer import ReLU, LeakyReLU, Sigmoid, ReLU6, Tanh, Softmax, PReLU
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: I prefer to write codes as follow (from its definition location), but you can choose it or not.

```python
from paddle.nn.layer.activation import ReLU, LeakyReLU, Sigmoid, ReLU6, Tanh, Softmax, PReLU
```

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 3]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
             layer, self._weight_bits, self._activation_bits, self._moving_rate,
             self._weight_quantize_type, self._activation_quantize_type)
         return quantized_layer
+
+
+class ImperativeOutScale(object):
+    def __init__(self,
+                 moving_rate=0.9,
+                 out_scale_layer_type=[
+                     'Conv2D', 'Pool2D', 'ReLU', 'BatchNorm', 'PRelu',
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Why not include `Linear` and `PReLU`?

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 4]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
...
+                     'Sigmoid', 'LeakyReLU', 'ReLU6', 'Tanh', 'Softmax',
+                     'Conv2DTranspose'
+                 ]):
+        """
+        Add the logic of calculating and setting output quantization scales of some layers.
+        These output quantization scales may be used by tensorRT or some other inference engines.
+
+        Args:
+            moving_rate(float): the parameter for 'moving_average_abs_max_scale' quantization.
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Use `The decay coefficient of moving average. The default value is 0.9.` instead.

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 5]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
...
+            quantizable_op_type(list[str]): List the type of layers that will be got out_scale. 
+                Default is ['Conv2D', 'ReLU', 'PReLU', 'LeakyReLU', 'Sigmoid', 'BatchNorm', 'ReLU6', 'Tanh', 'Softmax', 'Conv2DTranspose']
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Why not include `Linear` and `PReLU`?

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 6]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
...
+            'Pool2D': Pool2D,
+            'ReLU': ReLU,
+            'LeakyReLU': LeakyReLU,
+            'Sigmoid': Sigmoid,
+            'PRelu': PRelu,
+            'BatchNorm': BatchNorm,
+            'ReLU6': ReLU6,
+            'Tanh': Tanh,
+            'Softmax': Softmax,
+            'Conv2DTranspose': Conv2DTranspose
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Why not include `Linear` and `PReLU`?

File: python/paddle/fluid/contrib/slim/tests/test_imperative_out_scale.py [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+                bias_attr=fc_b2_attr),
+            Sigmoid(),
+            Linear(
+                input_dim=84,
+                output_dim=num_classes,
+                act=classifier_activation,
+                param_attr=fc_w3_attr,
+                bias_attr=fc_b3_attr))
+
+    def forward(self, inputs):
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/tests/test_imperative_out_scale.py:159]: Add `Softmax` test in UT.

File: python/paddle/fluid/contrib/slim/tests/test_imperative_out_scale.py [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+        seed = 1000
+        lr = 0.1
+        dynamic_out_scale_list = []
+        static_out_scale_list = []
+
+        # imperative train
+        _logger.info(
+            "--------------------------dynamic graph qat--------------------------"
+        )
+        imperative_out_scale = ImperativeOutScale()
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/tests/test_imperative_out_scale.py]: imperative_out_scale = ImperativeOutScale(target_layer_types = ['Conv2D', 'BatchNorm', 'ReLU', 'Pool2D', 'PRelu', 'Linear', 'LeakyReLU', 'Sigmoid', 'Softmax'])

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 9]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
             layer, self._weight_bits, self._activation_bits, self._moving_rate,
             self._weight_quantize_type, self._activation_quantize_type)
         return quantized_layer
+
+
+class ImperativeOutScale(object):
+    def __init__(self,
+                 moving_rate=0.9,
+                 out_scale_layer_type=[
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Replace `out_scale_layer_type ` with `target_layer_types`.

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 10]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
             layer, self._weight_bits, self._activation_bits, self._moving_rate,
             self._weight_quantize_type, self._activation_quantize_type)
         return quantized_layer
+
+
+class ImperativeOutScale(object):
+    def __init__(self,
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py:197]: Replace `ImperativeOutScale` with `ImperativeCalcOutScale`.

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 11]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
...
+        }
+        self._out_scale_layer_type = tuple(
+            self._out_scale_layers_map[layer]
+            if layer in self._out_scale_layers_map else layer
+            for layer in out_scale_layer_type)
+        for layer in self._out_scale_layer_type:
+            assert not isinstance(
+                layer, str), "{} is unspported to be out_scaled.".format(layer)
+
+    def get_out_scale(self, model):
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Replace `get_out_scale` with `calc_out_scale`.

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 12]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
...
+    def _forward_post_hook(self, layer, input, output):
+        if not isinstance(output, list):
+            output = [output]
+        scale_out_list = []
+        for var in output:
+            if var.dtype not in [
+                    core.VarDesc.VarType.FP32, core.VarDesc.VarType.FP64
+            ]:
+                continue
+            if not hasattr(layer, "_out_scale"):
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Why need this `if` statement ?

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 13]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
...
+                layer._out_scale = quant_nn.MovingAverageAbsMaxScale(
+                    var.name, self._moving_rate, var.dtype)
+            scale_out = layer._out_scale(var)
+            scale_out_list.append(float(scale_out.numpy()))
+        self._out_scale_dict[layer.full_name()] = scale_out_list
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Use the following codes instead.
```python
def _forward_post_hook(self, layer, input, output):
    if not isinstance(output, list):
        output = [output]
    out_scales = {}
    for var in output:
        if var.dtype not in [
                core.VarDesc.VarType.FP32, core.VarDesc.VarType.FP64
        ]:
            continue
        calc_out_scale = quant_nn.MovingAverageAbsMaxScale(
            var.name, self._moving_rate, var.dtype)
        scale_out = calc_out_scale(var)
        out_scales[var.name] = scale_out
    self._out_scale_dict[layer.full_name()] = out_scales
```

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 14]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
...
+        assert isinstance(
+            model, dygraph.Layer), "model must be the instance of dygraph.Layer"
+        for _, layer in model.named_sublayers():
+            if not isinstance(layer, self._out_scale_layer_type):
+                continue
+            scale_out_list = self._out_scale_dict[layer.full_name()]
+            for scale_var in scale_out_list:
+                layer.__setattr__('out_threshold', scale_var)
+        for handle in self._register_hook_handle_list:
+            handle.remove()
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Use the following codes instead.

```python
def set_out_scale(self, model):
    """
    Set the output quantization scale to the corresponding Layer of model.
    Args:
        model(fluid.dygraph.Layer): The output scale would be added to the model.
    Returns:
        None
    """
    assert isinstance(
        model, dygraph.Layer), "model must be the instance of dygraph.Layer"
    for _, layer in model.named_sublayers():
        if not isinstance(layer, self._out_scale_layer_type):
            continue
        out_scales = self._out_scale_dict[layer.full_name()]
        for name, scale in out_scales.items():
            # `name + '_out_threshold'` may not the proper name.
            layer.__setattr__(name + '_out_threshold', np.float32(scale.numpy()))
    for handle in self._register_hook_handle_list:
        handle.remove()
```

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 15]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
...
+        """
+        Insert the `moving_average_abs_max_scale` op to calculate output scale of Specific layers in model.
+
+        Args:
+            model(fluid.dygraph.Layer): The target model which would be calculate the output quantization scale.
+
+        Returns:
+            None
+        """
+        self._register_hook_handle_list = []
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Put the definition of `self._register_hook_handle_list` in `__init__`.

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 16]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
...
+        assert isinstance(
+            model, dygraph.Layer), "model must be the instance of dygraph.Layer"
+        self._out_scale_dict = {}
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: Put the definition of `self._out_scale_dict` in `__init__`.

**Review by wzzju — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 17]
```diff
@@ -237,3 +236,102 @@ def _get_quantized_counterpart(self, layer):
...
+            None
+        """
+        assert isinstance(
+            model, dygraph.Layer), "model must be the instance of dygraph.Layer"
+        for _, layer in model.named_sublayers():
+            if not isinstance(layer, self._out_scale_layer_type):
+                continue
+            scale_out_list = self._out_scale_dict[layer.full_name()]
+            for scale_var in scale_out_list:
+                layer.__setattr__('out_threshold', scale_var)
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py]: What do you want to do here? 🤒

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py (hunk 10 above)
Comment (gfwm2013) [reply python/paddle/fluid/contrib/slim/quantization/imperative/qat.py:197]: Thanks, done.

File: paddle/fluid/pybind/op_function_generator.cc (hunk 1 above)
Comment (gfwm2013) [reply paddle/fluid/pybind/op_function_generator.cc]: Thanks, it has been deleted.

File: python/paddle/fluid/contrib/slim/quantization/imperative/qat.py [hunk 18]
```diff
@@ -173,3 +189,176 @@ def _get_quantized_counterpart(self, layer):
...
+        These output quantization scales may be used by tensorRT or some other inference engines.
+
+        Args:
+            moving_rate(float): The decay coefficient of moving average. The default value is 0.9.
+            quantizable_op_type(list[str]): List the type of layers that will be calculated out_scale. 
+                Default is ['Conv2D', 'ReLU', 'PReLU', 'LeakyReLU', 'Linear', 'Sigmoid', 'BatchNorm', 'ReLU6', 'Tanh', 'Softmax', 'Conv2DTranspose']
+        """
+        super(ImperativeCalcOutScale, self).__init__()
+        self._moving_rate = moving_rate
+        self._out_scale_layers_map = {
```
Comment (juncaipeng) [inline python/paddle/fluid/contrib/slim/quantization/imperative/qat.py:215]: More ops should be supported.

**Review by juncaipeng — COMMENTED** [review]
收集输出scale的功能是量化训练的一部分，所以这部分api可以不用暴露给用户。
如果可以在ImperativeQuantAware中调用ImperativeCalcOutScale，可以更加简洁

**Review by zhiqiu — APPROVED** [review]
LGTM for op_function_generator.cc

File: python/paddle/fluid/contrib/slim/tests/test_imperative_out_scale.py [hunk 19]
```diff
@@ -0,0 +1,461 @@
...
+                avg_loss = fluid.layers.mean(loss)
+                avg_loss.backward()
+                adam.minimize(avg_loss)
+                lenet.clear_gradients()
+                dynamic_loss_rec.append(avg_loss.numpy()[0])
+                if batch_id % 100 == 0:
+                    _logger.info('{}: {}'.format('loss', avg_loss.numpy()))
+
+            lenet.eval()
+            op_object_list = (Conv2D, ReLU, ReLU6, LeakyReLU, Sigmoid, Pool2D,
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/tests/test_imperative_out_scale.py:343]: This is useless here.

File: python/paddle/fluid/contrib/slim/tests/test_imperative_out_scale.py [hunk 20]
```diff
@@ -0,0 +1,461 @@
...
+        seed = 1000
+        lr = 0.1
+        dynamic_out_scale_list = []
+        static_out_scale_list = []
+
+        # imperative train
+        _logger.info(
+            "--------------------------dynamic graph qat--------------------------"
+        )
+        imperative_out_scale = ImperativeCalcOutScale()
```
Comment (wzzju) [inline python/paddle/fluid/contrib/slim/tests/test_imperative_out_scale.py:298]: Use `target_layer_types` to create the object of `ImperativeCalcOutScale`.

**Review by wzzju — APPROVED** [review]
LGTM.
